### PR TITLE
lintcheck: support git sources

### PR DESCRIPTION
**Pull request by matthiaskrgr**
This adds support for git sources in `cargo dev-lintcheck`

You can add a git source to `clippy_dev/lintcheck_crates.toml` by having a `git_url` and a `git_hash` key instead of the `versions` array.
The repo will the be cloned and checked out to the requested commit before checking it with clippy.

Fixes https://github.com/rust-lang/rust-clippy/issues/6642

changelog: lintcheck: support git sources

### CODE REVIEW

File: clippy_dev/src/lintcheck.rs [hunk 1]
```diff
@@ -160,19 +201,40 @@ fn read_crates() -> Vec<CrateSource> {
...
+                url: tk.git_url.clone().unwrap(),
+                commit: tk.git_hash.clone().unwrap(),
             });
-        })
+        }
+        // if we have a version as well as a git data OR only one git data, something is funky
+        if tk.versions.is_some() && (tk.git_url.is_some() || tk.git_hash.is_some())
+            || tk.git_hash.is_some() != tk.git_url.is_some()
+        {
+            dbg!(&tk);
```
Comment (flip1995) [inline clippy_dev/src/lintcheck.rs:232]: dbg macro

File: clippy_dev/src/lintcheck.rs [hunk 2]
```diff
@@ -128,7 +164,12 @@ impl Crate {
             ])
             .current_dir(&self.path)
             .output()
-            .unwrap();
+            .unwrap_or_else(|error| {
+                dbg!(error);
+                dbg!(&cargo_clippy_path);
+                dbg!(&self.path);
```
Comment (flip1995) [inline clippy_dev/src/lintcheck.rs:170]: I don't think we should have dbg macros in our code.

**Review by flip1995 — APPROVED** [review]
With dbg macros removed.

File: clippy_dev/src/lintcheck.rs (hunk 1 above)
Comment (matthiaskrgr) [reply clippy_dev/src/lintcheck.rs:232]: I would have added some kind of error eprintln that prints all these but imo the `dbg` macro already prints all the information that I would want in the case of an error.

**Pull request merged**